syntax = "proto3";

package Blocksettle.Communication.Otc;

// Keep in sync with bs::network::otc::RangeType
enum RangeType
{
    RANGE_TYPE_0_1     = 0;
    RANGE_TYPE_1_5     = 1;
    RANGE_TYPE_5_10    = 2;
    RANGE_TYPE_10_50   = 3;
    RANGE_TYPE_50_100  = 4;
    RANGE_TYPE_100_250 = 5;
    RANGE_TYPE_250PLUS = 6;
}

// Keep in sync with bs::network::otc::Side
enum Side
{
    SIDE_INVALID = 0;
    SIDE_BUY = 1;
    SIDE_SELL = 2;
}

message Range
{
    int64 lower = 1;
    int64 upper = 2;
}

enum ContactType
{
    CONTACT_TYPE_PRIVATE = 0;
    CONTACT_TYPE_PUBLIC_REQUEST = 1;
    CONTACT_TYPE_PUBLIC_RESPONSE = 2;
}

message ContactMessage
{
    message Offer
    {
        int64 amount = 1;
        int64 price = 2;
    }

    message BuyerOffers
    {
        Offer offer = 1;

        // public key of auth address from buying side (33 bytes)
        bytes auth_address_buyer = 2;
    }

    message SellerOffers
    {
        Offer offer = 1;
    }

    message BuyerAccepts
    {
        Offer offer = 1;

        bytes auth_address_buyer = 2;
    }

    message SellerAccepts
    {
        // Check details here one more time
        Offer offer = 1;

        // 64 bytes generated by accepting side
        string settlement_id = 2;

        bytes auth_address_seller = 3;

        bytes payin_tx_id = 4;
    }

    message BuyerAcks
    {
        string settlement_id = 1;
    }

    message Close
    {
    }

    message QuoteResponse
    {
        Side sender_side = 1;
        Range amount = 2;
        Range price = 3;
    }

    ContactType contact_type = 10;

    oneof data
    {
        BuyerOffers buyer_offers = 1;
        SellerOffers seller_offers = 2;
        BuyerAccepts buyer_accepts = 3;
        SellerAccepts seller_accepts = 4;
        BuyerAcks buyer_acks = 5;
        Close close = 6;
        QuoteResponse quote_response = 7;
    }
}

message PublicMessage
{
    message Request
    {
        RangeType range = 1;
        Side sender_side = 2;
    }

    message Close
    {
    }

    oneof data
    {
        Request request = 1;
        Close close = 2;
    }
}
